Content.Server/Botany/Systems/MutationSystem.cs

```diff
@@ -0,0 +1,211 @@
+    //
+    // 'totalbits' and 'mult' are used only to calculate the probability that
+    // one bit gets flipped.
+    private static void MutateFloat(ref float val, float min, float max, int bits, int totalbits, float mult)
+    {
+        // Probability that a bit flip happens for this value.
+        float p = mult*bits/totalbits;
+        if (!random(p))
+        {
+            return;
+        }
+
+        // Starting number of bits that are high, between 0 and n.
+        int n = (int)Math.Round((val - min) / (max - min) * bits);
+
+        // Probability that the bit flip increases n.
+        float p_increase = 1-(float)n/bits;
+        int np;
+        if (random(p_increase))
+        {
+            np = n + 1;
+        }
+        else
+        {
+            np = n - 1;
+        }
+
+        // Set value based on mutated thermometer code.
+        float nval = (float)Math.Min(Math.Max((float)np/bits * (max - min) + min, min), max);
+        val = nval;
+    }
+
+    private static void MutateInt(ref int n, int min, int max, int bits, int totalbits, float mult)
+    {
+        // Probability that a bit flip happens for this value.
+        float p = mult*bits/totalbits;
+        if (!random(p))
+        {
+            return;
+        }
+
+        // Probability that the bit flip increases n.
+        float p_increase = 1-(float)n/bits;
+        int np;
+        if (random(p_increase))
+        {
+            np = n + 1;
+        }
+        else
+        {
+            np = n - 1;
+        }
+
+        np = Math.Min(Math.Max(np, min), max);
+        n = np;
+    }
+
+    private static void MutateBool(ref bool val, bool polarity, int bits, int totalbits, float mult)
+    {
+        // Probability that a bit flip happens for this value.
+        float p = mult*bits/totalbits;
+        if (!random(p))
+        {
+            return;
+        }
+
+        val = polarity;
+    }
+
+    private static Color RandomColor(Color color, int bits, int totalbits, float mult)
+    {
+        float p = mult*bits/totalbits;
+        if (random(p))
+        {
+            var colors = new List<Color>{
+                Color.White,
+                Color.Red,
+                Color.Yellow,
+                Color.Green,
+                Color.Blue,
+                Color.Purple,
+                Color.Pink
+            };
+            var rng = IoCManager.Resolve<IRobustRandom>();
+            return rng.Pick(colors);
+        }
+        return color;
+    }
+
+    private static void CrossFloat(ref float val, float other)
+    {
+        val = random(0.5f) ? val : other;
+    }
+
+    private static void CrossInt(ref int val, int other)
+    {
+        val = random(0.5f) ? val : other;
+    }
+
+    private static void CrossBool(ref bool val, bool other)
+    {
+        val = random(0.5f) ? val : other;
+    }
+
+    private static bool random(float p)
+    {
+        var rng = IoCManager.Resolve<IRobustRandom>();
```

Comment:
This should probably be a dependency rather than being resolved

Reply:
Sure.